protocol/app/upgrades/v5.0.0/upgrade.go

```diff
@@ -19,20 +20,69 @@ import (
 	"github.com/dydxprotocol/v4-chain/protocol/lib"
 )
 
-// Set all existing perpetuals to cross market type
+// perpetuals upgrade handler for v5.0.0
+// - 1. Set all perpetuals to cross market type
+// - 2. Initialize perpetual open interest to current OI
 func perpetualsUpgrade(
 	ctx sdk.Context,
 	perpetualsKeeper perptypes.PerpetualsKeeper,
+	subaccountsKeeper satypes.SubaccountsKeeper,
 ) {
-	// Set all perpetuals to cross market type
+	perpOIMap := make(map[uint32]*big.Int)
+
+	// Iterate through all subaccounts and perp positions for each subaccount.
+	// Aggregate open interest for each perpetual market.
+	subaccounts := subaccountsKeeper.GetAllSubaccount(ctx)
+	for _, sa := range subaccounts {
+		for _, perpPosition := range sa.PerpetualPositions {
+			if perpPosition.Quantums.BigInt().Sign() <= 0 {
+				// Only record positive positions for total open interest.
+				// Total negative position size should be equal to total positive position size.
+				continue
+			}
+			if openInterest, exists := perpOIMap[perpPosition.PerpetualId]; exists {
+				// Already seen this perpetual. Add to open interest.
+				openInterest.Add(
+					openInterest,
+					perpPosition.Quantums.BigInt(),
+				)
+			} else {
+				// Haven't seen this pereptual. Initialize open interest.
+				perpOIMap[perpPosition.PerpetualId] = new(big.Int).Set(
+					perpPosition.Quantums.BigInt(),
+				)
+			}
+		}
+	}
+
 	perpetuals := perpetualsKeeper.GetAllPerpetuals(ctx)
-	for _, p := range perpetuals {
-		_, err := perpetualsKeeper.SetPerpetualMarketType(
```

Comment:
Avoid two separate calls of `SetPerpetualMarketType` here and `ModifyOpenInterest` below; just use `validateAndSetPerpetual` once.